src/main-oe/sgxlkl_run_oe.c

```diff
@@ -504,30 +440,229 @@ static void sgxlkl_loader_signal_handler(int signo)
+    if (errno == EINVAL || errno == ERANGE)
+        sgxlkl_host_fail("Failed to parse hash offset!\n");
+
+    if (hashoffset_path != verity_file_or_hashoffset)
+        free(hashoffset_path);
+}
+
+static void override_enclave_disk_config(
+    sgxlkl_enclave_root_config_t* disk,
+    const char* image_path,
+    char* keyfile_or_passphrase,
+    char* verity_file_or_roothash,
+    char* verity_file_or_hashoffset)
+{
+    // If key and root hash are provided remotely or is set via enclave
+    // config, don't set it here.
+    if (host_state.enclave_config.mode != HW_RELEASE_MODE)
+    {
+        if (keyfile_or_passphrase)
+        {
+            // Currently we have a single parameter for passphrases and
+            // keyfiles. Determine which one it is.
+            if (access(keyfile_or_passphrase, R_OK) != -1)
+            {
+                FILE* kf;
+
+                if (!(kf = fopen(keyfile_or_passphrase, "rb")))
+                    sgxlkl_host_fail(
+                        "Failed to open keyfile %s.\n", keyfile_or_passphrase);
+
+                fseek(kf, 0, SEEK_END);
+                disk->key_len = ftell(kf);
+                if (disk->key_len > MAX_KEY_FILE_SIZE_KB * 1024)
+                {
+                    sgxlkl_host_warn(
+                        "Provided key file is larger than maximum supported "
+                        "key "
+                        "file size (%dkB). Only the first %dkB will be used.\n",
+                        MAX_KEY_FILE_SIZE_KB,
+                        MAX_KEY_FILE_SIZE_KB);
+                    disk->key_len = MAX_KEY_FILE_SIZE_KB * 1024;
+                }
+                rewind(kf);
+
+                disk->key = (uint8_t*)malloc((disk->key_len));
+                if (!fread(disk->key, disk->key_len, 1, kf))
+                    sgxlkl_host_fail(
+                        "Failed to read keyfile %s.\n", keyfile_or_passphrase);
+
+                fclose(kf);
+            }
+            else
+            {
+                disk->key_len = strlen(keyfile_or_passphrase);
+                disk->key = (uint8_t*)malloc(disk->key_len);
+                memcpy(disk->key, keyfile_or_passphrase, disk->key_len);
+            }
+        }
 
-    conf->app_config_str = buf;
+        prepare_verity(
+            disk,
+            image_path,
+            verity_file_or_roothash,
+            verity_file_or_hashoffset);
+    }
+}
+
+void override_disk_config(const char* root_disk_path)
+{
+    sgxlkl_enclave_config_t* config = &host_state.enclave_config;
+
+    /* Count disks to add */
+    const char* hds_str = sgxlkl_config_str(SGXLKL_HDS);
+    if (config->num_mounts == 0)
+    {
+        if (hds_str[0])
+        {
+            config->num_mounts++;
+            for (int i = 0; hds_str[i]; i++)
+            {
+                if (hds_str[i] == ',')
+                    config->num_mounts++;
+            }
+        }
+    }
+
+    if (!config->mounts)
```

Comment:
No, it can be empty when you run with an enclave config file which doesn't contain any mounts (=secondary disks) and you want to override that with mount settings from the corresponding environment variable.

Reply:
I still think it should be initialized elsewhere with an empty array. This doesn't seem like the right place here.

Reply:
The initialization happens in the default config (see line 1686: `host_state.enclave_config = sgxlkl_default_enclave_config;`) and the json-config may or may not overwrite that, so you may end up with the default (`NULL`) here. It might help to make that initialization-to-default-config more obvious by moving it into the other config-related functions (potentially duplicating it, but that may or may not be a problem).

Reply:
Hm, for arrays we shouldn't have a default of `NULL`  but instead an empty array. Is there a case where we need the additional semantics of no value vs empty array?

Reply:
In the JSON schema, it's an empty array (I added it explicitly now), but in the C struct all empty arrays are `NULL`.

Reply:
So why not change it? What's the benefit of NULL?

Reply:
Hmm, I'm afraid I don't understand the question. Do you want to `... = malloc(0);` it (which is usually also NULL)? This is C, so there's no fancy `std::vector<T>`. (And no, `[]` is not a viable solution either.)

Reply:
Hm, you're right, I totally forgot about that. Nevermind...

Reply:
No worries, happens to me too sometimes :-)